From 4e79520c927f3e3c5b95306fba4ee04553cd493c Mon Sep 17 00:00:00 2001
From: Paul Chavent <paul.chavent@onera.fr>
Date: Fri, 28 Sep 2012 17:56:37 +0200
Subject: [PATCH 1/3] Lua : fix corba deployer

If TaskContextServer::Create or TaskContextServer::ThreadOrb fails, it isn't
useful to continue with a CorbaDeploymentComponent, so fallback to a DeploymentComponent.

Restore the -- separator for lua in this case.

Use Deployer instead of deployer as name for the deployer so it's compatible with the
legacy deployer.
---
 lua/LuaComponent.cpp |   28 +++++++++++++++++++++-------
 1 file changed, 21 insertions(+), 7 deletions(-)

diff --git a/lua/LuaComponent.cpp b/lua/LuaComponent.cpp
index 0a56197..3e913fb 100644
--- lua/LuaComponent.cpp
+++ lua/LuaComponent.cpp
@@ -302,14 +302,17 @@ int ORO_main(int argc, char** argv)
 #endif
 
 	LuaComponent lua("lua");
-	DeploymentComponent * dc;
+	DeploymentComponent * dc = 0;
 
 #if defined(LUA_RTT_CORBA)
 	int  orb_argc = argc;
 	char** orb_argv = 0;
+	char* orb_sep = 0;
         
+	/* find the "--" separator */
 	while(orb_argc) {
 	  if(0 == strcmp("--", argv[argc - orb_argc])) {
+	    orb_sep = argv[argc - orb_argc];
 	    argv[argc - orb_argc] = argv[0];
 	    orb_argv = &argv[argc - orb_argc];
 	    argc -= orb_argc;
@@ -318,11 +321,12 @@ int ORO_main(int argc, char** argv)
 	  orb_argc--;
 	}
         
+	/* if the "--" separator is found perhaps we have orb arguments */
 	if(orb_argc) {
 	  try {
 	    TaskContextServer::InitOrb(orb_argc, orb_argv);
    
-	    dc = new CorbaDeploymentComponent("deployer");
+	    dc = new CorbaDeploymentComponent("Deployer");
    
 	    TaskContextServer::Create( dc, true, true );
    
@@ -332,17 +336,27 @@ int ORO_main(int argc, char** argv)
 	  catch( CORBA::Exception &e ) {
 	    log(Error) << argv[0] <<" ORO_main : CORBA exception raised!" << Logger::nl;
 	    log() << CORBA_EXCEPTION_INFO(e) << endlog();
+	    if(dc)
+	    {
+	      delete dc;
+	      dc = 0;
+	    }
 	  } catch (...) {
 	    log(Error) << "Uncaught exception." << endlog();
+	    if(dc)
+	    {
+	      delete dc;
+	      dc = 0;
+	    }
 	  }
  
-	  // lua_repl doesn't use argc ?
-	  argv[argc] = NULL;
+	  argv[argc] = dc?NULL:orb_sep;
 	}
-	else
-#endif
 
-	dc = new DeploymentComponent("deployer");
+	/* fallback to the default deployer if corba have failed to provide one */
+	if(!dc)
+#endif
+	dc = new DeploymentComponent("Deployer");
 
 	lua.connectPeers(dc);
 
